### PR TITLE
LibWeb+LibGfx: Add support for scalable and theme-able checkboxes using SDFs     

### PR DESCRIPTION
**Pull request by MacDue**
This PR adds new smoothly scaling check boxes (using a simple singed distance field),  that respect the system them and/or the `accent-color` CSS property.

Check boxes are just a start here, I think this method could be used to add more scalable UI in other places too, without too much extra complexity over bitmaps :^)

|                                                                                                                 | Testing all themes                                                                                                               |
|-----------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------|
|image | ezgif com-video-to-gif (1) |

**Smooth scaling demo:**

[screen-capture - 2023-03-20T001833.748.webm](https://user-images.githubusercontent.com/11597044/226219497-019273b6-b0eb-428a-af08-dc33fe9a2773.webm)

### CODE REVIEW

**Comment by skyrising** [comment]
I'm not sure SDFs are the right tool for the job here. You're kind of forced to make the check mark have rounded corners and it's encoded in an opaque list of numbers.

My suggestion for scalable check boxes would be to put an SVG into the <input>'s shadow root and wire the accent-color to the SVG's fill color.

**Comment by awesomekling** [comment]
IMO it would be even nicer to do this in the `Gfx::StylePainter` so we can get nice scaled checks when using SerenityOS with larger UI scale as well :)

**Comment by MacDue** [comment]
@skyrising While I agree using vector graphics would be very nice, the anti-aliasing painter is not at the point where it can render something like this with anywhere near the quality you get from a SDF. And I did spend quite a long time trying before switching to the SDF approach, but at small scales (e.g. 1x the icons become illegible, and things don't scale smoothly).

There's many issues with precision, antialiasing, and painting which means it just does not look too good right now (that's not to say I don't want to keep trying to improve the AA painter, I love the AA painter, it's just a lot more of a complexity than simple SDFs).

**Comment by MacDue** [comment]
@awesomekling I think that'd be very neat, but I'm not sure if it would look out of place until a few more inputs if are made in the same style :^) (I can test that this evening). Since I was going for a more modern web look for the LibWeb one. Though it could be possible to mess around with the box style and smoothing to get one that looks similar to current one, but with more scaling possible?

**Comment by AtkinsSJ** [comment]
I just merged https://github.com/SerenityOS/serenity/pull/17940 which created a bit of a conflict, sorry! I thought it better to merge that one first, and maybe you can do similar SDF magic for the indeterminate state.

**Comment by MacDue** [comment]
No worries, I'll do that when I have some time this evening :)

**Comment by MacDue** [comment]
(Added improved support for dark themes, see updated gif)

File: Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp [hunk 1]
```diff
@@ -33,4 +90,72 @@ Layout::CheckBox& CheckBoxPaintable::layout_box()
...
+            if (color.contrast_ratio(background) < min_contrast) {
+                color = color.inverted();
+                if (color.contrast_ratio(background) > min_contrast)
+                    return color;
+            }
+            return color;
+        };
+
+        float smoothness = 1.0f / (max(checkbox_rect.width(), checkbox_rect.height()) / 2);
+        if (checkbox.checked()) {
```
Comment (srikavin) [inline Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp:138]: One small thing. The indeterminate state should take precedence over checked: 
> If the element's [indeterminate](https://html.spec.whatwg.org/multipage/input.html#dom-input-indeterminate) IDL attribute is set to true, then the control's selection should be obscured as if the control was in a third, indeterminate, state.

[spec](https://html.spec.whatwg.org/multipage/input.html#checkbox-state-(type=checkbox):dom-input-indeterminate)

Comment (MacDue) [reply Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp:138]: Ah, thank you!

File: Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp [hunk 2]
```diff
@@ -33,4 +90,72 @@ Layout::CheckBox& CheckBoxPaintable::layout_box()
...
 }
 
+void CheckBoxPaintable::paint(PaintContext& context, PaintPhase phase) const
+{
+    if (!is_visible())
+        return;
+
+    PaintableBox::paint(context, phase);
+
+    if (phase == PaintPhase::Foreground) {
```
Comment (GMTA) [inline Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp:100]:
```suggestion
    if (phase != PaintPhase::Foreground)
        return;
```

File: Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp [hunk 3]
```diff
@@ -33,4 +90,72 @@ Layout::CheckBox& CheckBoxPaintable::layout_box()
...
+        auto const& checkbox = static_cast<HTML::HTMLInputElement const&>(layout_box().dom_node());
+        bool enabled = layout_box().dom_node().enabled();
+        Gfx::AntiAliasingPainter painter { context.painter() };
+        auto checkbox_rect = context.enclosing_device_rect(absolute_rect()).to_type<int>();
+        auto checkbox_radius = checkbox_rect.width() / 5;
+
+        auto& palette = context.palette();
+
+        auto shade = [&](Color color, float amount = 0.3f) {
```
Comment (GMTA) [inline Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp:109]: It feels arbitrary to have a default `amount = 0.3f` when you specify the amount almost everywhere else. Why not always set the required `amount` in the calling code?

File: AK/StdLibExtras.h [hunk 4]
```diff
@@ -119,6 +119,13 @@ constexpr T mix(T const& v1, T const& v2, U const& interpolation)
     return v1 + (v2 - v1) * interpolation;
 }
 
+template<typename T, typename U>
+constexpr T smooth_step(T edge0, T edge1, U x)
+{
+    x = clamp((x - edge0) / (edge1 - edge0), 0.0f, 1.0f);
+    return x * x * (3 - 2 * x);
+}
```
Comment (GMTA) [inline AK/StdLibExtras.h:127]: As far as I know, "the" smooth step function is a primarily graphics (shader) driven function that isn't commonly part of any standard library that I know of. Incorporating this in LibGfx might be fine though?

**Review by GMTA — COMMENTED** [review]
Always nice to see an SDF implementation! I'm not sure if it makes sense to use SDF for something that is basically vector graphics, and if so, if we should incorporate the distance fields in the binaries or if we should load them from external files - but someone more LibWeb-minded might say something about that.

**Comment by MacDue** [comment]
I previously mentioned that currently the AA painter is not at the point where it can smoothly render and scale icons (while maintaining legibility across a range of scales), but the SDFs produce a really nice and smoothly scaling result without much extra complexity (see https://github.com/SerenityOS/serenity/pull/17933#issuecomment-1475857390).

Embedding them in the binary is done because these are super small and it removes bunch of complexity for loading them in Ladybird and such.

File: AK/StdLibExtras.h (hunk 4 above)
Comment (MacDue) [reply AK/StdLibExtras.h:127]: I placed it there because we have other functions that feel similar there (like `mix()` and `clamp()`, and such)

**Comment by GMTA** [comment]
@MacDue 

> I previously mentioned [..]

I know, I'm just noting that I'm not sure whether we should improve the AA painter or incorporate SDFs. If people are all for it, then why not :^)

**Comment by MacDue** [comment]
We should improve the AA painter! (and it's something I'm trying to do too!) But getting something as general as SVG painting to match the quality of the SDFs is much bigger challenge, and I think SDFs make for a nice solution now :^)

File: AK/StdLibExtras.h (hunk 4 above)
Comment (GMTA) [reply AK/StdLibExtras.h:127]: Things like interpolating/mixing and clamping are much more ubiquitous and definitely not limited to graphics, whereas I've never seen smoothstep outside of graphics processing.

File: Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp (hunk 3 above)
Comment (MacDue) [reply Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp]: No real reason (other than `0.3f` felt like a nice default increment between shades, but I'll remove it).

File: AK/StdLibExtras.h (hunk 4 above)
Comment (MacDue) [reply AK/StdLibExtras.h:127]: I'll just move this to be a lambda where it's used for now (and we can find a home for if it's ever needed elsewhere).

**Comment by konradekk** [comment]
(not sure if itʼs any how important but iʼd reorder the 4-frame images to be bit accessible, i.e. first two being enabled and latter two disabled, in those groups first could be ticked and the other unticked [or other way around]. havenʼt thought how iʼd would order tri-state, though. please ignore this comment if not about something which might be of any use or interest in the future!)

File: Userland/Libraries/LibGfx/GrayscaleBitmap.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Gfx {
+
+class GrayscaleBitmap {
+public:
+    GrayscaleBitmap() = delete;
+    constexpr GrayscaleBitmap(ReadonlyBytes data, unsigned width, unsigned height)
+        : m_data(data)
+        , m_size(width, height)
+    {
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGfx/GrayscaleBitmap.h:22]: :thinking: Can we `VERIFY(width * height == data.size());`?

File: Userland/Libraries/LibWeb/CSS/Default.css [hunk 6]
```diff
@@ -64,33 +64,6 @@ button:hover, input[type=submit]:hover, input[type=button]:hover, input[type=res
     background-color: -libweb-palette-hover-highlight;
 }
 
-input[type=checkbox] {
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Default.css:67]: Commit message: I don't know what "animalised" means but the definitions I've found don't make sense here so I'm assuming it's a typo?

**Review by AtkinsSJ — COMMENTED** [review]
A couple of small things. Very cool!

File: Userland/Libraries/LibGfx/GrayscaleBitmap.h (hunk 5 above)
Comment (MacDue) [reply Userland/Libraries/LibGfx/GrayscaleBitmap.h:22]: Yep :+1:

**Comment by MacDue** [comment]
@konradekk I've re-ordered the demo to be a bit more sensible now (if that's what you meant) :)
image

**Review by AtkinsSJ — APPROVED** [review]
:white_check_mark: :ballot_box_with_check: :heavy_check_mark:

**Pull request merged**